### ISSUE TITLE
alternative to indicator allpairs if nconditions too large

### ISSUE THREAD

**Issue by iancharest** [rsagroup/rsatoolbox#37]
when dealing with a large number of conditions, e.g. with the natural scenes dataset, indicator will fail with a memory issue.

should we have a fallback plan to compute diff_1, and diff_2 the slow (but memory friendly) way if measurements1.shape[0] > something ?

```
c_matrix = allpairs(np.arange(measurements1.shape[0]))
diff_1 = np.matmul(c_matrix, measurements1)
diff_2 = np.matmul(c_matrix, measurements2)
```

**Comment by HeikoSchuett** [comment]
I want to bump this up, because I currently look at implementing the basic demo and for 92 stimuli the matrix based calculation is very slow!
according to the profiler the main culprit is the generation of the matrix

**Comment by HeikoSchuett** [comment]
yes, indeed:
```
from pyrsa.util.matrix import pairwise_contrast
pairwise_contrast(np.arange(92))
```
takes 1.5s on my computer

**Comment by iancharest** [comment]
speed is an issue, but i think memory is an even bigger problem, try 
```python
pairwise_contrast(np.arange(1000))
```

**Comment by iancharest** [comment]
one quick fix would be to declare indicator_matrix as type int 16. are we ever expecting anything else than zeros, ones and -1s?

**Comment by HeikoSchuett** [comment]
well as it is coded now it could be fractions when there are multiple measurements for the same condition. Also this would only save a factor 2 or 4 compared to the float32/64 right?

With my change the 1000 contrast matrix can actually be calculated on my computer.
It is very slow and memory expensive still and we should probably switch to either a sparse matrix or calculating this entirely without representing the matrix as you originally suggested here.

**Comment by iancharest** [comment]
ok, but by now by default this will be a float64, which is probably an overkill. float32 would make it half the size and should be plenty? (the difference between 3Gb or 1.5Gb in RAM is not small.
I wonder what the actual gain in speed is compared to doing the difference in a loop?

**Comment by HeikoSchuett** [comment]
I guess that a loop is faster than the current implementation, at least for large numbers of conditions.
Overall I think: We need a way to do this better. And I think there should be ways to save much more than a factor of 2. For your example case this matrix has the size of 1000 rdms which should definitely be avoidable.

**Comment by HeikoSchuett** [comment]
I now put in my sparse matrix implementation. 
This takes a few seconds to calculate a 1000x1000 rdm from a 1000x1000 dataset. Is that acceptable?

**Comment by HeikoSchuett** [comment]
closed by #49
